visibility:hidden fails to hide composited iframe

rwlbuis opened the pull request:
#### 172c0c9a864d40d004786767fe4ba00cd7e97fb6
<pre>
visibility:hidden fails to hide composited iframe
<a href="https://bugs.webkit.org/show_bug.cgi?id=220342">https://bugs.webkit.org/show_bug.cgi?id=220342</a>

Reviewed by NOBODY (OOPS!).

Turn off compositing for subframes when they are hidden through
visibility:hidden.

* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-ancestor-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-ancestor.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-toggled-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-toggled.html: Added.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingLayer const):
(WebCore::RenderLayerCompositor::isHiddenSubframe const):
(WebCore::RenderLayerCompositor::requiresCompositingForFrame const):
* Source/WebCore/rendering/RenderLayerCompositor.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/172c0c9a864d40d004786767fe4ba00cd7e97fb6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5939 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7333 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6167 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5775 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6161 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7943 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5880 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5927 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5219 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7387 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3404 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5203 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13038 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7434 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5725 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4698 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5169 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5133 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9287 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5530 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -2642,6 +2642,9 @@ bool RenderLayerCompositor::requiresCompositingLayer(const RenderLayer& layer, R
         return false;
     }
 
+    if (isHiddenSubframe(renderer))
```
This is the wrong place for this; it should be in `requiresCompositingForFrame()`.

rwlbuis replied:
The idea to place it here is to ignore any other reasons for composting when dealing with hidden subframes, like requiresCompositingForTransform which would be true in the tests and would override requiresCompositingForFrame.

smfr commented on Source/WebCore/rendering/RenderLayerCompositor.cpp:
```diff
@@ -3328,16 +3331,22 @@ bool RenderLayerCompositor::requiresCompositingForPlugin(RenderLayerModelObject&
 
     auto& frameRenderer = downcast<RenderWidget>(renderer);
-    if (frameRenderer.style().visibility() != Visibility::Visible)
+    return frameRenderer.style().visibility() != Visibility::Visible;
```
Does an iframe inside a `visibility:hidden` ancestor get composited? Is there code somewhere else that prevents that?

We should also test a `visibility:visible` iframe in a `visibility:hidden` ancestor.

rwlbuis replied:
I added the test (composited-iframe-visibility-hidden-ancestor.html).

rwlbuis committed:
visibility:hidden fails to hide composited iframe
https://bugs.webkit.org/show_bug.cgi?id=220342

Reviewed by NOBODY (OOPS!).

Turn off compositing for subframes when they are hidden through
visibility:hidden.

* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-ancestor-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-ancestor.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-hidden.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-toggled-expected.html: Added.
* LayoutTests/compositing/iframes/composited-iframe-visibility-toggled.html: Added.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingLayer const):
(WebCore::RenderLayerCompositor::isHiddenSubframe const):
(WebCore::RenderLayerCompositor::requiresCompositingForFrame const):
* Source/WebCore/rendering/RenderLayerCompositor.h:
